### PR TITLE
Add websockets support

### PR DESCRIPTION
**Pull request by alex-schneider**
Due to our `endpoint` feature for multiple requests and proxies websocket upgrade requests were not supported since 1.0.
This PR introduces a `websockets` option for the `proxy` block to enable and configure those upgrade requests.

---
<details>
    <summary>Reviewer checklist</summary>
    <ul>
        <li>Read PR description: a summary about the changes is required</li>
        <li>Changelog updated</li>
        <li>Documentation: docs/{Reference, Cli, ...}, Docker and cli help/usage</li>
        <li>Pulled branch, manually tested</li>
        <li>Verified requirements are met</li>
        <li>Reviewed the code</li>
        <li>Reviewed the related tests</li>
    </ul>
</details>

### CODE REVIEW

File: eval/context.go [hunk 1]
```diff
@@ -210,8 +210,10 @@ func (c *Context) WithBeresps(beresps ...*http.Response) *Context {
 		}.Merge(newVariable(ctx.inner, bereq.Cookies(), bereq.Header)))
 
 		var respBody, respJsonBody cty.Value
-		if (ctx.bufferOption & BufferResponse) == BufferResponse {
-			respBody, respJsonBody = parseRespBody(beresp)
+		if !CheckUpgradeResponse(bereq, beresp) {
```
[the remark this answers is not in the document]
Comment (malud) [reply eval/context.go:213]: Also a comment line would be nice within this context to note why reading the body should be skipped.

File: handler/proxy.go [hunk 2]
```diff
@@ -62,7 +66,14 @@ func (p *Proxy) RoundTrip(req *http.Request) (*http.Response, error) {
 		*req = *req.WithContext(ctx)
 	}
 
-	rec := transport.NewRecorder()
+	rw := req.Context().Value(request.RW).(*writer.Response)
+	rec := transport.NewRecorder(rw)
+
+	if p.allowWebsockets {
```
Comment (malud) [inline handler/proxy.go:72]: As already said: dynamic value

**Review by malud — CHANGES_REQUESTED** [review]

File: config/websockets.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+// Schema implements the <Inline> interface.
+func (w Websockets) Schema(inline bool) *hcl.BodySchema {
+	schema, _ := gohcl.ImpliedBodySchema(w)
+	if !inline {
+		return schema
+	}
+
+	type Inline struct {
+		SetRequestHeaders map[string]string `hcl:"set_request_headers,optional"`
```
Comment (malud) [inline config/websockets.go:31]: What about `set_response_headers` ? Could get registered via responseWriter modifier.

File: docs/REFERENCE.md [hunk 4]
```diff
@@ -278,6 +279,19 @@ The `oauth2` block in the [Backend Block](#backend-block) context configures the
 | `token_endpoint_auth_method` |string|`client_secret_basic`|Defines the method to authenticate the client at the token endpoint.|If set to `client_secret_post`, the client credentials are transported in the request body. If set to `client_secret_basic`, the client credentials are transported via Basic Authentication.|-|
 | `scope`                      |string|-|  A space separated list of requested scopes for the access token.|-| `scope = "read write"` |
 
+### Websockets Block
+
+The `websockets` block activates support for websocket connection in Couper.
```
Comment (malud) [inline docs/REFERENCE.md:284]:
```suggestion
The `websockets` block activates support for websocket connections in Couper.
```

File: handler/proxy.go [hunk 5]
```diff
@@ -53,6 +58,46 @@ func (p *Proxy) RoundTrip(req *http.Request) (*http.Response, error) {
 		return nil, err
 	}
 
+	content, _, diags := p.context.PartialContent(config.Proxy{Remain: p.context}.Schema(true))
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	if wss := content.Blocks.OfType("websockets"); len(wss) == 1 {
```
Comment (malud) [inline handler/proxy.go:66]: extract this block into a method.

File: config/configload/load.go [hunk 6]
```diff
@@ -470,6 +470,24 @@ func refineEndpoints(definedBackends Backends, endpoints config.Endpoints, check
...
+			wsEnabled, wsBody, wsErr := getWebsocketsConfig(proxyConfig)
+			if wsErr != nil {
+				return wsErr
+			}
+
+			if wsEnabled {
+				if proxyConfig.Name != defaultNameLabel {
+					return errors.Configuration.Message("websockets attribute or block is only allowed in a 'default' proxy block")
+				}
+				if proxyRequestLabelRequired || endpoint.Response != nil {
```
[the remark this answers is not in the document]
Comment (malud) [reply config/configload/load.go:482]: I recommend to review https://github.com/avenga/couper/pull/294 beforehand.

**Review by malud — CHANGES_REQUESTED** [review]

File: config/request/context_key.go [hunk 7]
```diff
@@ -5,19 +5,22 @@ type ContextKey uint8
 const (
 	UID ContextKey = iota
 	AccessControls
+	AllowWebsockets
```
[the remark this answers is not in the document]
Comment (alex-schneider) [reply config/request/context_key.go:8]: We can do it later if necessary...

File: handler/proxy.go [hunk 8]
```diff
@@ -86,6 +97,51 @@ func (e *ErrorWrapper) Write(p []byte) (n int, err error) {
...
 	return log.New(&ErrorWrapper{logger}, "", log.Lshortfile)
 }
+
+func (p *Proxy) applyWebsockets(req *http.Request) error {
+	bodyContent, _, diags := p.context.PartialContent(config.Proxy{Remain: p.context}.Schema(true))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	if wss := bodyContent.Blocks.OfType("websockets"); len(wss) == 1 {
```
Comment (malud) [inline handler/proxy.go:111]:
```suggestion
	wss := bodyContent.Blocks.OfType("websockets")
	if len(wss) != 1 {
		return nil
	}
```

reduce method complexity

File: handler/proxy.go [hunk 9]
```diff
@@ -86,6 +97,51 @@ func (e *ErrorWrapper) Write(p []byte) (n int, err error) {
...
+		ctx := req.Context()
+
+		ctx = context.WithValue(ctx, request.AllowWebsockets, true)
+		*req = *req.WithContext(ctx)
+
+		// This method needs the 'request.AllowWebsockets' flag in the 'req.context'.
+		if eval.IsUpgradeRequest(req) {
```
Comment (malud) [inline handler/proxy.go:118]: also early out

File: handler/proxy.go [hunk 10]
```diff
@@ -86,6 +97,51 @@ func (e *ErrorWrapper) Write(p []byte) (n int, err error) {
...
+			bodyContent, _, diags = wss[0].Body.PartialContent(config.WebsocketsInlineSchema)
+			if diags.HasErrors() {
+				return diags
+			}
+			if err := eval.ApplyRequestContext(req.Context(), wss[0].Body, req); err != nil {
+				return err
+			}
+
+			if attr, ok := bodyContent.Attributes["timeout"]; ok {
```
Comment (malud) [inline handler/proxy.go:127]: also early out, without timeout just return `nil`.

File: config/request/context_key.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (malud) [reply config/request/context_key.go:8]: Take a look at the existing ones. We have a convention already:

Endpoint*
RoundTrip*
TokenRequest*

so a good fit would be
WebsocketsAllowed/Enabled
WebsocketsTimeout

**Review by malud — CHANGES_REQUESTED** [review]
Some things:

- [x] applyWebsockets method complexity can be reduced
- [ ] `set_response_headers` within `websockets` does not work at all
- [x] missing test for `set_response_headers`
- [x] apply context order must be: proxy body, proxy blacklist, websockets body
- [x] without order fixup, `set_request_headers = { cookie = request.headers.cookie }` does not work

File: server/writer/response.go [hunk 11]
```diff
@@ -130,14 +142,28 @@ func (r *Response) WriteHeader(statusCode int) {
 	r.configureHeader()
 	r.applyModifier()
 
+	writeStatusCode := statusCode
```
Comment (alex-schneider) [inline server/writer/response.go:149]: Why this new var?

File: server/writer/response.go [hunk 12]
```diff
@@ -106,12 +111,10 @@ func (r *Response) Hijack() (net.Conn, *bufio.ReadWriter, error) {
...
 
-	conn, rw, err := hijack.Hijack()
-	if err != nil {
-		return nil, nil, err
-	}
-
-	return conn, rw, err
+	conn, brw, err := hijack.Hijack()
+	r.hijackedConn = conn
+	brw.Writer.Reset(r)
```
Comment (alex-schneider) [inline server/writer/response.go:116]: panic() candidate if err != nil

**Review by malud — APPROVED** [review]

**Pull request merged**